scapy/contrib/pim.py

```diff
@@ -0,0 +1,252 @@
+# This file is part of Scapy
+# Scapy is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# any later version.
+#
+# Scapy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Scapy. If not, see <http://www.gnu.org/licenses/>.
+#
+# scapy.contrib.description = Protocol Independent Multicast (PIM)
+# scapy.contrib.status = loads
+"""
+References:
+  - https://tools.ietf.org/html/rfc4601
+  - https://www.iana.org/assignments/pim-parameters/pim-parameters.xhtml#pim-parameters-1  # noqa: E501
+"""
+import struct
+from scapy.packet import Packet, bind_layers
+from scapy.fields import BitFieldLenField, BitField, BitEnumField, ByteField, \
+    ShortField, XShortField, IPField, PacketListField, \
+    IntField, FieldLenField, BoundStrLenField, FlagsField
+from scapy.layers.inet import IP
+from scapy.utils import checksum
+from scapy.compat import orb
+from scapy.config import conf
+from scapy.volatile import RandInt
+
+
+PIM_TYPE = {
+    0: "Hello",
+    1: "Register",
+    2: "Register-Stop",
+    3: "Join/Prune",
+    4: "Bootstrap",
+    5: "Assert",
+    6: "Graft",
+    7: "Graft-Ack",
+    8: "Candidate-RP-Advertisement"
+}
+
+
+class PIMv2Hdr(Packet):
+    name = "Protocol Independent Multicast Version 2 Header"
+    fields_desc = [BitField("version", 2, 4),
+                   BitEnumField("type", 0, 4, PIM_TYPE),
+                   ByteField("reserved", 0),
+                   XShortField("chksum", None)]
+
+    def post_build(self, p, pay):
+        """
+        Called implicitly before a packet is sent to compute and place PIM checksum.  # noqa: E501
```

Comment:
Please avoid using noqa tags for new code.

Reply:
I apply feedback.

https://github.com/secdev/scapy/pull/2757/commits/0ebd9a2ced53f8bb8ea43d2bdd0abd743a6af46c